openhtf/exe/__init__.py

```diff
@@ -122,40 +137,99 @@ def _RunTestLoop(self):
       # Top level steps required to run a single iteration of the Test.
       _LOG.info('Starting test %s', self.test.code_info.name)
 
-      # Save exit_stack on self so we can access it when Stop() is called.
-      self._current_exit_stack = exit_stack
-      exit_stack.callback(lambda: setattr(self, '_current_exit_stack', None))
-
+      # Any access to self._exit_stack must be done while holding this lock.
+      with self._lock:
+        # Initial setup of exit stack and final cleanup of attributes.
+        self._exit_stack = exit_stack
+        exit_stack.callback(self._ResetAttributes)
+        exit_stack.callback(self._MaybeJoinOutputThread)
+
+      # Wait here until the test start trigger returns a DUT ID.  Don't hold
+      # self._lock while we do this, or else calls to Stop() will deadlock.
+      # Create plugs while we're here because that may also take a while and
+      # we don't want to hold self._lock while we wait.
+      dut_id = self._WaitForTestStart(suppressor)
+      self._status = self.FrameworkStatus.INITIALIZING
       plug_manager = self._MakePlugManager(suppressor)
-      executor = self._MakePhaseExecutor()
 
+      with self._lock:
+        if not self._exit_stack:
+          # We shouldn't get here, but just in case something went weird with a
+          # call to Stop() and we ended up resuming execution here but the
+          # exit stack was already cleared, bail.  Try to tear down plugs on a
+          # best-effort basis.
+          plug_manager.TearDownPlugs()
+          raise TestStopError('Test Stopped.')
+        exit_stack.callback(plug_manager.TearDownPlugs)
+
+        # Perform initialization of some top-level stuff we need.
+        self._test_state = self._MakeTestState(dut_id, plug_manager, suppressor)
+        executor = self._MakePhaseExecutor(exit_stack, suppressor)
+
+      # Everything is set, set status and begin test execution.  Note we don't
+      # protect this with a try: block because the PhaseExecutor handles any
+      # exceptions from test code.  Any exceptions here are caused by the
+      # framework, and we probably want them to interrupt framework state
+      # changes (like the transition to FINISHING).
       self._status = self.FrameworkStatus.EXECUTING
       suppressor.failure_reason = 'Failed to execute test.'
       self._ExecuteTestPhases(executor)
       self._status = self.FrameworkStatus.FINISHING
-      self.test.OutputTestRecord(self._test_state.GetFinishedRecord())
 
-  def _MakePlugManager(self, suppressor):
-    """Perform some initialization and create a PlugManager."""
+      # Output the test record now that we're done executing.
+      self._output_thread = self._MakeAndStartOutputThread()
```

Comment:
We'll have to toss some idea around.  I haven't spent too much time thinking about the best way to implement it, but I'd like to add some more safety around output callbacks (right now they can raise and it's awkward) and provide some kind of fallback mechanism.

The use for this is mainly "upload the result somewhere, and if it fails save it locally to disk".  The problem with "try to upload and always also save to disk" is that depending on the upload API, it may be hard to check "has this one already been uploaded".  Anyway we can chat about that later, going to make those changes and get this in :)